### ISSUE TITLE
Typing information is lost when a variable is captured within lambda

### ISSUE THREAD

**Issue by povilasb** [microsoft/pyright#2157]
**Describe the bug**
1. See the code example to reproduce the issue.
2. The exact problem is that a type info is lost when we capture a lambda argument within another lambda:
```py
Flow[None]().then(lambda res: [1, 2, 3]).then(
    lambda numbers: lambda: reveal_type(numbers)
)

info: Type of "numbers" is "Unknown"
```
3. The type of the first lambda is correctly recognized:
```py
Flow[None]().then(lambda res: [1, 2, 3]).then(lambda numbers: reveal_type(numbers))

Type of "numbers" is "list[int]"
```

4. The problem seems to be with overloads. If I remove overloads and leave it like this:
```py
from typing import Callable, Generic, overload, TypeVar, cast

_OutT = TypeVar("_OutT")
_Out2T = TypeVar("_Out2T")


class Flow(Generic[_OutT]):
    def then(self: "Flow[_OutT]", func: Callable[[_OutT], _Out2T]) -> "Flow[_Out2T]":
        return cast("Flow[_Out2T]", self)


Flow[None]().then(lambda res: [1, 2, 3]).then(
    lambda numbers: lambda: reveal_type(numbers)
)

Type of "numbers" is "list[int]"
```
It works fine.

**To Reproduce**
```py
from typing import Callable, Generic, overload, TypeVar

_OutT = TypeVar("_OutT")
_Out2T = TypeVar("_Out2T")


class Flow(Generic[_OutT]):
    @overload
    def then(self: "Flow[_OutT]", func: Callable[[_OutT], int]) -> "Flow[str]":
        ...

    @overload
    def then(self: "Flow[_OutT]", func: Callable[[_OutT], _Out2T]) -> "Flow[_Out2T]":
        ...

    def then(
        self,
        func: Callable,
    ) -> "Flow":
        return self


Flow[None]().then(lambda res: [1, 2, 3]).then(lambda numbers: lambda: numbers)
```

**VS Code extension or command-line**

Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

I can reproduce the same behavior on VS code and `pyright` CLI tool:
```bash
pyright --version
pyright 1.1.160

pylance v2021.7.7
```

This is a very specific scenario and hope I'm missing something :)
Let me know if you need more info.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.161, which I just published. It will also be included in the next release of pylance.

**Comment by povilasb** [comment]
Thank you thousand times!
